### PR TITLE
noobaa operator crd checks 

### PR DESCRIPTION
**Pull request by kerencarmelwedel**
noobaa crd checks

### CODE REVIEW

File: test/cli/test_cli_functions.sh [hunk 1]
```diff
@@ -284,6 +284,23 @@ function obc_cycle {
     echo_time "✅  obc cycle is done"
 }
 
+function crd_cycle {  
+    local crd 
+    echo_time "Starting the crd cycle"
+    test_noobaa timeout crd create
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:290]: running `noobaa crd create` when all crd exist will result in:

```
INFO[0000] ✅ Already Exists: CustomResourceDefinition "noobaas.noobaa.io"
INFO[0000] ✅ Already Exists: CustomResourceDefinition "backingstores.noobaa.io"
INFO[0000] ✅ Already Exists: CustomResourceDefinition "bucketclasses.noobaa.io"
INFO[0000] ✅ Already Exists: CustomResourceDefinition "objectbucketclaims.objectbucket.io"
INFO[0000] ✅ Already Exists: CustomResourceDefinition "objectbuckets.objectbucket.io"
```

in order to check that we need delete then create and check that it exists, this flow will test the delete command and the create command

File: test/cli/test_cli_functions.sh [hunk 2]
```diff
@@ -284,6 +284,23 @@ function obc_cycle {
...
+    test_noobaa timeout crd wait
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:291]: what is it testing?

File: test/cli/test_cli_functions.sh [hunk 3]
```diff
@@ -284,6 +284,23 @@ function obc_cycle {
...
+    crd_array=($(kubectl get crd | awk '{print $1}' | grep -v "NAME"))
+    echo_time "${crd_array[*]}"
+    for crd in ${crd_array[@]}
+    do 
+        test_noobaa timeout crd status
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:296]: `noobaa crd status` returns all the crds, we don't need it inside the loop

we should check that each card expected is present here, after we do status

File: test/cli/test_cli_functions.sh [hunk 4]
```diff
@@ -284,6 +284,23 @@ function obc_cycle {
...
+        kuberun describe crd ${crd}
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:297]: running this command alone will not do. 
What do we want to test?

File: test/cli/test_cli_functions.sh [hunk 5]
```diff
@@ -284,6 +284,23 @@ function obc_cycle {
...
+    test_noobaa timeout crd create
+    test_noobaa timeout crd wait
+    crd_array=($(kubectl get crd | awk '{print $1}' | grep -v "NAME"))
+    echo_time "${crd_array[*]}"
+    for crd in ${crd_array[@]}
+    do 
+        test_noobaa timeout crd status
+        kuberun describe crd ${crd}     
+    done
+    kuberun get crd
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:299]: This command output is:
```
$ kuberun get crd
NAME                                 CREATED AT
backingstores.noobaa.io              2020-05-13T08:15:27Z
bucketclasses.noobaa.io              2020-05-13T08:15:27Z
noobaas.noobaa.io                    2020-05-13T08:15:27Z
objectbucketclaims.objectbucket.io   2020-05-13T08:15:27Z
objectbuckets.objectbucket.io        2020-05-13T08:15:27Z
11:20:22 ✅  kubectl get crd passed
```
or when there is no crd:
```
$ kuberun get crd
No resources found in default namespace.
11:51:07 ✅  kubectl get crd passed
```

we should verify that what we are expecting is there, and run this after each command (create, delete)

File: test/cli/test_cli_functions.sh [hunk 6]
```diff
@@ -284,6 +284,66 @@ function obc_cycle {
     echo_time "✅  obc cycle is done"
 }
 
+function crd_cycle {  
+    local crd 
+    echo_time "Starting the crd cycle" 
+    
+    test_noobaa timeout crd create
+    noobaa crd status &>/dev/stdout | grep -v "Already Exists"
```
Comment (nimrod-becker) [inline test/cli/test_cli_functions.sh:292]:
```suggestion
    test+noobaa crd status &>/dev/stdout | grep -v "Already Exists"  
```

**Comment by guymguym** [comment]
@kerencarmelwedel I labelled this PR as "Work in progress".
Remove this label when you're ready to review/merge. Thanks

File: test/cli/test_cli_functions.sh [hunk 7]
```diff
@@ -120,12 +120,11 @@ function timeout {
 
 function install {
     test_noobaa install --mini
-    local status=$(kuberun get noobaa noobaa -o json | jq -r '.status.phase' 2> /dev/null)
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:123]: If we remove this line better replace it with `local status` to set this variable as local

File: test/cli/test_cli_functions.sh [hunk 8]
```diff
@@ -228,13 +227,13 @@ function bucketclass_cycle {
...
         backingstore+=("compatible$((number+1))")
-    done
+     done
 
     test_noobaa bucketclass create ${bucketclass_names[0]} --backingstores ${backingstore[0]}
     # test_noobaa bucketclass create ${bucketclass_names[1]} --placement Mirror --backingstores nb1,aws1 ❌
     # test_noobaa bucketclass create ${bucketclass_names[2]} --placement Spread --backingstores aws1,aws2 ❌
-    test_noobaa bucketclass create ${bucketclass_names[3]} --backingstores ${backingstore[0]},${backingstore[1]}
-
+     test_noobaa bucketclass create ${bucketclass_names[3]} --backingstores ${backingstore[0]},${backingstore[1]}
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:235]:
```suggestion
    test_noobaa bucketclass create ${bucketclass_names[3]} --backingstores ${backingstore[0]},${backingstore[1]}
```

File: test/cli/test_cli_functions.sh [hunk 9]
```diff
@@ -228,13 +227,13 @@ function bucketclass_cycle {
...
+
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:236]:
```suggestion

```

File: test/cli/test_cli_functions.sh [hunk 10]
```diff
@@ -284,10 +283,97 @@ function obc_cycle {
...
+     else 
+        echo_time "✅ crd already exsists in the test"   
+    fi  
+  
+    echo_time "checking if crd noobaa create command equals to crd status command:"
+    for crd_status_array in ${crd_array[@]}
+    do
+        if [[ ${crd_create_array[@]} =~ ${crd_status_array} ]]
+        then            
+            echo_time "✅ ${crd_status_array} is exsisted in the crds list"
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:306]:
```suggestion
            echo_time "✅ ${crd_status_array} is exist in the crds list"
```

File: test/cli/test_cli_functions.sh [hunk 11]
```diff
@@ -268,7 +267,7 @@ function obc_cycle {
         then
             flag="--app-namespace default"
         fi
-        test_noobaa timeout obc create ${buckets[$((${#buckets[@]}-1))]} --bucketclass ${bucketclass} ${flag}
+        #test_noobaa timeout obc create ${buckets[$((${#buckets[@]}-1))]} --bucketclass ${bucketclass} ${flag}
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:270]: Why is it in comment?

File: test/cli/test_cli_functions.sh [hunk 12]
```diff
@@ -284,10 +283,97 @@ function obc_cycle {
     echo_time "✅  obc cycle is done"
 }
 
+function crd_cycle {  
+    crd_array=($(kubectl get crd | awk '{print $1}' | grep -v "NAME"))
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:287]: Why isn't it local?

File: test/cli/test_cli_functions.sh [hunk 13]
```diff
@@ -284,10 +283,97 @@ function obc_cycle {
...
+    local crd_create_array=($(test_noobaa silence crd create &>/dev/stdout | awk '{print $7}' | grep -v "Already Exists" | sed -e 's/[""\\]//g' | sed 's/.$//'))
+    local crd_status_array=($(test_noobaa silence crd status &>/dev/stdout | awk '{print $6}' | grep -v "Exists" | sed -e 's/[""\\]//g' | sed 's/.$//'))
+    echo_time "Starting the crd cycle"
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:290]: This print should be at the start of the function.

File: test/cli/test_cli_functions.sh [hunk 14]
```diff
@@ -284,10 +283,97 @@ function obc_cycle {
...
+    if [ $? -ne 0 ]  
+    then           
+        echo_time "❌ crd doesn't exist in the test. exiting."
+        exit 1 
+     else 
+        echo_time "✅ crd already exsists in the test"   
+    fi  
+  
+    echo_time "checking if crd noobaa create command equals to crd status command:"
+    for crd_status_array in ${crd_array[@]}
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:302]: Why this loop is here? if we are not populating the arrays here, put this loop next to it.
If we want it after the second create, then we need to populate the array with its output.

File: test/cli/test_cli_functions.sh [hunk 15]
```diff
@@ -284,10 +283,97 @@ function obc_cycle {
...
+        fi               
+    done
+         
+    kuberun get crd  
+    if [ $? -ne 0 ]
+    then
+        echo_time "❌ kubectl get crd failed"
+        exit 1 
+    else 
+        echo_time "✅kubectl get crd passed"
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:318]:
```suggestion
        echo_time "✅  kubectl get crd passed"    
```

File: test/cli/test_cli_functions.sh [hunk 16]
```diff
@@ -284,10 +283,97 @@ function obc_cycle {
...
+    for name in ${crd_array[@]} 
+    do
+        if [[ ${crd_status_array[@]} =~ ${crd_status_after_delete} ]]
+        then    
+            echo_time "❌ crd ${name} still exists in the test. exiting."
+            exit 1       
+        else 
+            echo_time "✅ crd ${name} deleted from test" 
+        fi
+    done
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:334]: We want to check that a crd in the crd_array not presents in the crd_status_after_delete array
```suggestion
    for crd in ${crd_array[@]} 
    do
        if [[ ${crd_status_after_delete[@]} =~ ${crd} ]]
        then    
            echo_time "❌ crd ${name} still exists in the test. exiting."
            exit 1       
        else 
            echo_time "✅ crd ${name} deleted from test" 
        fi
    done
```

File: test/cli/test_cli_functions.sh [hunk 17]
```diff
@@ -284,10 +283,97 @@ function obc_cycle {
...
+    else 
+        echo_time "✅ kubectl get crd passed"    
+    fi 
+
+    echo_time "creating crd again for checks..."     
+    test_noobaa silence crd create 
+    
+    for crd_status_array in ${crd_array[@]}
+    do
+        if [[ ${crd_create_array[@]} =~ ${crd_status_array} ]]
```
Comment (liranmauda) [inline test/cli/test_cli_functions.sh:349]: those arrays are populated at the beginning of the function, We need to repopulate.

File: test/cli/test_cli_functions.sh (hunk 7 above)
Comment (kerencarmelwedel) [reply test/cli/test_cli_functions.sh:123]: re-added

File: test/cli/test_cli_functions.sh (hunk 12 above)
Comment (kerencarmelwedel) [reply test/cli/test_cli_functions.sh:288]: Also used in another function - uninstall with cleanup

**Comment by kerencarmelwedel** [comment]
closed pr, created a new one #334

**Pull request closed without merging**